lib/iris/fileformats/netcdf.py

```diff
@@ -1958,7 +1949,11 @@ def set_packing_ncattrs(cfvar):
             # Explicitly assign the fill_value, which will be the type default
             # in the case of an unmasked array.
             if packing is None:
-                fill_value = cube.fill_value
+                if hasattr(cube.lazy_data(), 'fill_value'):
+                    fill_value = cube.lazy_data().fill_value
+                else:
+                    fill_value = None
+
```

Comment:
Yeah, that's behaviour that we just can't avoid, and it's concerning. If we're saving lazy masked data, then the fill-value is lost. Not good.

We could do a work around though ... we know that we can only get the fill-value from concrete or non-lazy masked data, so in the lazy masked case, we could slice the lazy data down to 1 data item, then compute it, then get the fill-value. That works, and it's similar to a trick we've used in netcdf to calculate the derived dtype from data that requires a scale + offset calculation.

As you mentioned, we could also apply this approach to PP to ensure that we don't lose the goodness of @pp-mo's #2452.

@lbdreyer How does that sound?